### ISSUE TITLE
Starting neo4j server with an existing local database dir out of installation path fails

### ISSUE THREAD

**Issue by dalanlan** [neo4j/neo4j#7155]
### Steps to reproduce
1. install `$ brew install neo4j`
2. import data from legacy databse via `neo4j-import`, say the graph path is my_graph_absolute_path
3. edit `/usr/local/Cellar/neo4j/3.0.0/libexec/conf/neo4j.conf`:  #`dbms.active_database=my_graph_absolute_path`
4. try to connect the graph db and see what we got so far `curl --user neo4j:pass_word http://localhost:7474/db/data/node/1` and tada what a shame

```
...
"message" : "Node 1 not found",
    "errors" : [ {
      "code" : "Neo.ClientError.Statement.EntityNotFound",
      "message" : "Node 1 not found"
    } ]
  },
  "errors" : [ {
    "code" : "Neo.ClientError.Statement.EntityNotFound",
    "message" : "Cannot find node with id [1] in database."

```

5 however, this is definitely not expected because I could double-check with `neo4j-shell`

```
neo4j-sh (1)$ ls
:Movie
*TMSId =[MV000000020000] 
(me)<-[:ACTED_IN]-(Steven  Hill,172637)
(me)<-[:ACTED_IN]-(Kenneth  McMillan,171072)
(me)<-[:ACTED_IN]-(William  Hurt,163079)
(me)<-[:ACTED_IN]-(Pamela  Reed,159909)
(me)<-[:ACTED_IN]-(Morgan  Freeman,159717)
(me)<-[:ACTED_IN]-(James  Woods,158002)
(me)<-[:ACTED_IN]-(Irene  Worth,157824)
(me)<-[:ACTED_IN]-(Christopher  Plummer,156353)
(me)<-[:ACTED_IN]-(Sigourney  Weaver,155286)
```
### Platform
- Neo4j version: 3.0.0
- Operating system: OS X 10.11.3
### My word

This could be a naive question(I've already put it on SO as you suggest yet with no luck), yet from my POV, I bet it should be specified more precisely. I would love to volunteer myself to that once this is settled.

**Comment by spacecowboy** [comment]
@dalanlan I'm not sure I can see a problem here. By importing the legacy database, the nodes are simply keeping whatever ID they already have?

Could you verify that the legacy database actually contains a Node with ID 1 prior to the import? If not, I'd say this is entirely expected.

**Comment by dalanlan** [comment]
@spacecowboy 
You could see the content of node 1 from the my original post, aka

```
neo4j-sh (2)$ cd -a 1
neo4j-sh (1)$ ls
:Movie
*TMSId =[MV000000020000] 
(me)<-[:ACTED_IN]-(Steven  Hill,172637)
(me)<-[:ACTED_IN]-(Kenneth  McMillan,171072)
(me)<-[:ACTED_IN]-(William  Hurt,163079)
(me)<-[:ACTED_IN]-(Pamela  Reed,159909)
(me)<-[:ACTED_IN]-(Morgan  Freeman,159717)
(me)<-[:ACTED_IN]-(James  Woods,158002)
(me)<-[:ACTED_IN]-(Irene  Worth,157824)
(me)<-[:ACTED_IN]-(Christopher  Plummer,156353)
(me)<-[:ACTED_IN]-(Sigourney  Weaver,155286)
```

**Comment by dalanlan** [comment]
More precisely, I cannot get a legacy database online with simply editing the conf field, if that's what I'm supposed to do.

**Comment by spacecowboy** [comment]
Can you post the log? Have you enabled [dbms.allow_format_migration](http://neo4j.com/docs/operations-manual/current/#config_dbms.allow_format_migration)?

**Comment by dalanlan** [comment]
@spacecowboy  
I dont believe I need to enable `dbms.allow_format_migration` 'cause they are of the same version (actually I should have never call it a legacy db).
What I've done is using `neo4j-import` to load data from mysql:)
I cannot find anything interesting in `/usr/local/Cellar/neo4j/3.0.0/libexec/logs/neo4j.log` though. Is there any other logs I might miss?

```
2016-05-17 04:50:24.639+0000 INFO  Starting...
2016-05-17 04:50:25.253+0000 INFO  Bolt enabled on localhost:7687.
2016-05-17 04:50:27.621+0000 INFO  Started.
2016-05-17 04:50:28.659+0000 INFO  Remote interface available at http://localhost:7474/
```

**Comment by spacecowboy** [comment]
@dalanlan Oh, when you said legacy I was sure you meant 2.3. So, to clear matters up then.. 
1. You have an existing database, and you export it as CSV files.
2. You import it using `neoj4j-import` as explained above, with the setting mentioned
3. Via the rest interface, there is no node with id 1
4. Via the shell (on the same running database as in 3) there IS a node with id 1

To summarize, the rest interface does not seem to read data from the same directory as the shell does?

There is also the debug log in `/usr/local/Cellar/neo4j/3.0.0/libexec/logs/debug.log` which you might want to have a look at.

**Comment by dalanlan** [comment]
> 1.You have an existing database, and you export it as CSV files.
> 2.You import it using neoj4j-import as explained above, with the setting mentioned
> 3.Via the rest interface, there is no node with id 1
> 4.Via the shell (on the same running database as in 3) there IS a node with id 1

Yes that's very right. You could save my day!
Any suggestions?

**Comment by spacecowboy** [comment]
ah thank you for the report. this does seem like a problem!

**Comment by dalanlan** [comment]
Meaning no one has gone through this routine before? I was thinking it's a very common use case. 
So how do people handle large amount of data migration then :( 

I only chose to use `neo4j-import` because `LOAD CSV` via cypher runs like a tutle (no offense, it could be my own problem). Like I could only insert 1000 relationships with about 10 min.

**Comment by benbc** [comment]
@dalanlan `dbms.directories.data` should not be the path of the database, it is the path of the `data` directory. It's likely that you do not need to set it at all.

Can you show me how you invoked `neo4j-import`?

**Comment by dalanlan** [comment]
@benbc 

```
$ neo4j-import --into graceLight.db --nodes:Movie absolute_movie_path.csv --nodes:Actor absolute_actor_path.csv --relationships:ACTED_IN absolute_member_path.csv --delimiter "|"
```

And I set `dbms.directories.data=absolute_path/graceLight.db`
What am I supposed to do then:)

**Comment by benbc** [comment]
@dalanlan Please try this:
1. Comment out the `dbms.directories.data` line in `neo4j.conf`.
2. Set `dbms.active_database=graceLight.db` in `neo4j.conf`.
3. Execute the following

```
$ cd <neo4j-installation-root>
$ bin/neo4j stop
$ neo4j-import --into data/databases/graceLight.db --nodes:Movie absolute_movie_path.csv --nodes:Actor absolute_actor_path.csv --relationships:ACTED_IN absolute_member_path.csv --delimiter "|"
$ bin/neo4j start
```

I think that will import the data correctly for you.

**Comment by dalanlan** [comment]
@benbc 

> Actually I've set `dbms.active_database=graceLight.db` before and it didn't work out. I guess that should be an absolute path instead?
> However the `db_path` in `neo4j-import --into db_path` is a relative path and I could not see why it works that way.

(Edit) I guess I can see your point. Not working though. 
However, is it possible that I could start a server from an existing data  dir and of course I don't want to bind with the installation root dir of neo4j:)

**Comment by dalanlan** [comment]
@benbc 
TL; DR
1. import data: `neo4j-import --into graceLight.db ...` (relative path)
2. set up path: `dbms.active_databas=absolute_path/graceLight.db`
3. `neo4j start` and give me nothing 
4. While the server is under process at step 3, I tried `neo4j-shell` and it failed to start, meaning the server is already started and seems I've got to the right place.One question only, where's the data then?

```
$ proof for step 4
...
Caused by: org.neo4j.kernel.StoreLockException: Store and its lock file has been locked by another process: absolute_path/graceLight.db/store_lock. Please ensure no other process is using this database, and that the directory is writable (required even for read-only access)
    at org.neo4j.kernel.internal.StoreLocker.storeLockException(StoreLocker.java:90)
    at org.neo4j.kernel.internal.StoreLocker.checkLock(StoreLocker.java:76)
    at org.neo4j.kernel.internal.StoreLockerLifecycleAdapter.start(StoreLockerLifecycleAdapter.java:40)
    at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:434)
    ... 14 more
```

Any idea? That would be much appreciated.

**Comment by dalanlan** [comment]
![screen shot 2016-05-19 at 10 11 25 am](https://cloud.githubusercontent.com/assets/5979704/15380568/1ef0c710-1daa-11e6-8c50-822f0e520fe7.png)
My browser is not responsive as well

**Comment by rishabhjain08** [comment]
@dalanlan I was able to run the imported data both in the shell and in the browser using the steps mentioned by @benbc . I think you should not set dbms.active_database to an absolute path because its used in conjunction with data.directories.data as a result the resultant path in your case will be malformed. (Though this still does not explain why it is working for shell but not for browser)
Can you try running neo4j-import so that  data/databases/graphLight.db (or simply move graphLight.db to data/databases/ from whereever it was created) is created.

**Comment by dalanlan** [comment]
@rishabhjain08 
I followed his guide and this is what I've got
```
$ neo4j-import --into data/databases/graceLight.db --nodes:Movie /Users/simei.he/Documents/graphDB/gracenote/movieTmp.csv --nodes:Actor /Users/simei.he/Documents/graphDB/gracenote/actorTmp.csv --relationships:ACTED_IN /Users/simei.he/Documents/graphDB/gracenote/memberTmp.csv --delimiter "|"
Input error: Directory '/usr/local/Cellar/neo4j/3.0.0/libexec/data/databases/graceLight.db' already contains a database
Caused by:Directory '/usr/local/Cellar/neo4j/3.0.0/libexec/data/databases/graceLight.db' already contains a database
java.lang.IllegalArgumentException: Directory '/usr/local/Cellar/neo4j/3.0.0/libexec/data/databases/graceLight.db' already contains a database
	at org.neo4j.kernel.impl.util.Validators$3.validate(Validators.java:100)
	at org.neo4j.kernel.impl.util.Validators$3.validate(Validators.java:93)
	at org.neo4j.helpers.Args.validated(Args.java:640)
	at org.neo4j.helpers.Args.interpretOption(Args.java:548)
	at org.neo4j.tooling.ImportTool.main(ImportTool.java:325)
	at org.neo4j.tooling.ImportTool.main(ImportTool.java:291)
```

**Comment by dalanlan** [comment]
Yet still, I was thinking it's such a pain that we must bind ourselves to the installation dir and could not find a way out.
Hence the issue remains open as it is.

**Comment by rishabhjain08** [comment]
You can still point to a database directory outside of the installation dir but you have to make sure that the path reads **data.directories.data**/databases/**dbms.active_database**
You can change **data.directories.data** to any directory outside of the installation dir. I hope that is what you mean.

**Comment by dalanlan** [comment]
@rishabhjain08  
That's very helpful and thank you so much. Do you think it should be specified more precisely in the document or did I just miss it myself?

**Comment by rishabhjain08** [comment]
Its not explicitly mentioned that databases should exist under data dir but I think there are enough hints to infer that that is the case. For eg. Under link http://neo4j.com/docs/operations-manual/current/#config_dbms.active_database heading "Upgrade from 3.x" bullet 5 and bullet 6 can be used to conclude this.

**Comment by dalanlan** [comment]
Hmm I bet I must be a moron if I think otherwise. Anyway thanks for your time.
My browser is still dying though. I'm gonna leave it as it is. 
Closing the issue.

**Comment by benbc** [comment]
@dalanlan I'm glad you've got this sorted out. Sorry that the behaviour isn't clear from the docs. @rishabhjain08 Thank you for helping to explain the problem.

To be completely clear, `dbms.active_database` is the _name_ of the database, not the path. To configure Neo4j to store data outside the installation root, you need to set a value for `dbms.directories.data`; that directory will then contain all data managed by Neo4j (database and some internal data such as authentication info). As of 3.0 it is not possible to configure paths for individual databases.

If either of you have any suggestions about how we could improve the documentation to avoid this kind of confusion in future, we would love to hear them.

**Comment by dalanlan** [comment]
@benbc More one thing, any chance we could avoid the middle `databases` in `data.directories.data/databases/dbms.active_database`?
